### PR TITLE
Switch location

### PR DESCRIPTION
**Pull request by jeevansurendran**
Everything other than the main page where the location needs to be shown is implemented i guess.

### CODE REVIEW

File: app/src/main/java/com/checkin/app/checkin/utility/Constants.kt [hunk 1]
```diff
@@ -30,7 +30,8 @@ object Constants {
     const val SP_MANAGER_LIVE_ORDER_ACTIVE_KEY = "com.checkin.app.checkin.sp.manager.live.order.key"
     const val SP_SHOP_PREFERENCES_TABLE = "com.checkin.app.checkin.sp.shop.preferences"
 
-    const val LOCATION_CITY_ID = "home.user.location.switch"
+    const val LOCATION_CITY_ID = "home.user.location.id"
+    const val LOCATION_CITY_FILE = "home.user.location"
```
Comment (jeevansurendran) [inline app/src/main/java/com/checkin/app/checkin/utility/Constants.kt]: Used these constant values for shared preferences

File: app/src/main/java/com/checkin/app/checkin/home/epoxy/CityLocationModelHolder.kt [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+import com.checkin.app.checkin.misc.epoxy.BaseEpoxyHolder
+
+@EpoxyModelClass(layout = R.layout.item_user_location_city)
+abstract class CityLocationModelHolder : EpoxyModelWithHolder<CityLocationModelHolder.Holder>() {
+
+    @EpoxyAttribute
+    internal lateinit var data: CityLocationModel
+
+    @EpoxyAttribute
+    internal lateinit var cityListener: View.OnClickListener
```
Comment (shivanshs9) [inline app/src/main/java/com/checkin/app/checkin/home/epoxy/CityLocationModelHolder.kt]: You should not use view click listener, but rather create a new interface. That's the standard way for the communication between different modules.

File: app/src/main/java/com/checkin/app/checkin/home/epoxy/CurrentLocationModelHolder.kt [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+import com.airbnb.epoxy.EpoxyModelClass
+import com.airbnb.epoxy.EpoxyModelWithHolder
+import com.checkin.app.checkin.R
+import com.checkin.app.checkin.misc.epoxy.BaseEpoxyHolder
+
+@EpoxyModelClass(layout = R.layout.item_user_location_current)
+abstract class CurrentLocationModelHolder : EpoxyModelWithHolder<CurrentLocationModelHolder.Holder>() {
+
+    @EpoxyAttribute
+    internal lateinit var currentListener: View.OnClickListener
```
Comment (shivanshs9) [inline app/src/main/java/com/checkin/app/checkin/home/epoxy/CurrentLocationModelHolder.kt]: Same here, you can use the same interface but pass default arguments to the function

File: app/src/main/java/com/checkin/app/checkin/home/fragments/UserLocationFragment.kt [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+
+
+            viewModel.locationData.value?.data?.forEachIndexed { index, model ->
+                cityLocationModelHolder {
+                    id(model.id)
+                    data(model)
+                    cityListener { _ ->
+                        val preferences = requireActivity().getSharedPreferences(Constants.LOCATION_CITY_FILE, Context.MODE_PRIVATE)
+                        with(preferences.edit()) {
+                            putInt(Constants.LOCATION_CITY_ID, model.id)
```
Comment (shivanshs9) [inline app/src/main/java/com/checkin/app/checkin/home/fragments/UserLocationFragment.kt]: The coding pattern is really wrong here.
You're using a model variable bound to the one in loop. It's better to use the interface I mentioned above and declare the parameters for city data

**Comment by jeevansurendran** [comment]
Current location shouldn't be shown when typed

File: app/src/main/java/com/checkin/app/checkin/home/fragments/UserLocationFragment.kt (hunk 4 above)
Comment (shivanshs9) [reply app/src/main/java/com/checkin/app/checkin/home/fragments/UserLocationFragment.kt]: You still didn't change that. I already mentioned to properly use interface and use the parameters in the interface

File: app/src/main/java/com/checkin/app/checkin/menu/listeners/LocationSelectedListener.java [hunk 5]
```diff
@@ -0,0 +1,5 @@
+package com.checkin.app.checkin.menu.listeners;
+
+public interface LocationSelectedListener {
```
Comment (shivanshs9) [inline app/src/main/java/com/checkin/app/checkin/menu/listeners/LocationSelectedListener.java]: Uh... Why the heck would you write this interface in menu package?

File: app/src/main/java/com/checkin/app/checkin/menu/listeners/LocationSelectedListener.java [hunk 6]
```diff
@@ -0,0 +1,5 @@
...
+    void onSelectedLocation();
```
Comment (shivanshs9) [inline app/src/main/java/com/checkin/app/checkin/menu/listeners/LocationSelectedListener.java]: And why the heck is this function without any parameters? I already told you that the location data will be sent as arguments to this function.

File: app/src/main/java/com/checkin/app/checkin/home/epoxy/CityLocationModelHolder.kt [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+import com.checkin.app.checkin.misc.epoxy.BaseEpoxyHolder
+
+@EpoxyModelClass(layout = R.layout.item_user_location_city)
+abstract class CityLocationModelHolder : EpoxyModelWithHolder<CityLocationModelHolder.Holder>() {
+
+    @EpoxyAttribute
+    internal lateinit var data: CityLocationModel
+
+    @EpoxyAttribute
+    internal lateinit var locationSelected: LocationSelectedListener
```
Comment (shivanshs9) [inline app/src/main/java/com/checkin/app/checkin/home/epoxy/CityLocationModelHolder.kt]: Field name is wrong - told you to use something similar to 'listener'

File: app/src/main/java/com/checkin/app/checkin/home/fragments/UserLocationFragment.kt [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+import com.checkin.app.checkin.misc.fragments.BaseFragment
+import com.checkin.app.checkin.utility.Constants
+
+class UserLocationFragment : BaseFragment() {
+    override val rootLayout = R.layout.fragment_user_location_switch
+    val viewModel: UserLocationViewModel by viewModels()
+    val homeViewModel: HomeViewModel by activityViewModels()
+    @BindView(R.id.et_user_location)
+    internal lateinit var etUserLocation: EditText
+
```
Comment (shivanshs9) [inline app/src/main/java/com/checkin/app/checkin/home/fragments/UserLocationFragment.kt]: Don't waste lines in between View fields... Fix that in all the files (like holder)

File: app/src/main/java/com/checkin/app/checkin/home/fragments/UserLocationFragment.kt [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+import com.checkin.app.checkin.home.epoxy.cityLocationModelHolder
+import com.checkin.app.checkin.home.epoxy.currentLocationModelHolder
+import com.checkin.app.checkin.home.viewmodels.HomeViewModel
+import com.checkin.app.checkin.home.viewmodels.UserLocationViewModel
+import com.checkin.app.checkin.misc.fragments.BaseFragment
+import com.checkin.app.checkin.utility.Constants
+
+class UserLocationFragment : BaseFragment() {
+    override val rootLayout = R.layout.fragment_user_location_switch
+    val viewModel: UserLocationViewModel by viewModels()
```
Comment (shivanshs9) [inline app/src/main/java/com/checkin/app/checkin/home/fragments/UserLocationFragment.kt]: Declare these variables after all the UI views are declared. Always keep rootLayout at the top, though.

File: app/src/main/java/com/checkin/app/checkin/home/epoxy/CityLocationModelHolder.kt [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+
+    override fun bind(holder: Holder) {
+        holder.bindData(data)
+        holder.itemView.setOnClickListener {
```
Comment (shivanshs9) [inline app/src/main/java/com/checkin/app/checkin/home/epoxy/CityLocationModelHolder.kt]: You sure you checked other existing holders in the app? Never ever touch UI elements outside the Holder class. If you want to set click listener, go do that by overriding the 'bindView' method in the Holder classes.

Comment (shivanshs9) [reply app/src/main/java/com/checkin/app/checkin/home/epoxy/CityLocationModelHolder.kt]: bind purpose is just to bind data. Not deal with UI elements.

File: app/src/main/java/com/checkin/app/checkin/home/epoxy/CityLocationModelHolder.kt [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+    }
+
+    class Holder : BaseEpoxyHolder<CityLocationModel>() {
+
+        @BindView(R.id.tv_user_location_city)
+        internal lateinit var tvCityLocation: TextView
+
+        @BindView(R.id.tv_user_location_state)
+        internal lateinit var tvStateLocation: TextView
+
```
Comment (shivanshs9) [inline app/src/main/java/com/checkin/app/checkin/home/epoxy/CityLocationModelHolder.kt]: Lots of needless blank lines here too. Always try and keep your code compact.

File: app/src/main/java/com/checkin/app/checkin/home/fragments/UserLocationFragment.kt [hunk 12]
```diff
@@ -0,0 +1,120 @@
...
+    @BindView(R.id.epoxy_rv_user_location)
+    internal lateinit var epoxyUserLocation: EpoxyRecyclerView
+
+    @BindView(R.id.im_user_location_cross)
+    internal lateinit var imUserCross: ImageView
+
+    @OnTextChanged(R.id.et_user_location, callback = OnTextChanged.Callback.AFTER_TEXT_CHANGED)
+    fun onTextChanged(et: Editable?) {
+        if (et.toString().isEmpty()) {
+            imUserCross.visibility = View.INVISIBLE
```
[the remark this answers is not in the document]
Comment (shivanshs9) [reply app/src/main/java/com/checkin/app/checkin/home/fragments/UserLocationFragment.kt]: It still isn't resolved.

**Pull request merged**